gleam_stdlib/src/list.gleam

```diff
@@ -54,15 +54,19 @@ pub fn filter(list, fun) {
   do_filter(list, fun, [])
 }
 
-fn do_map(list, fun, acc) {
+fn do_map_with(list, f, acc, map_f) {
   case list {
   | [] -> reverse(acc)
-  | [x | xs] -> do_map(xs, fun, [fun(x) | acc])
+  | [x | xs] -> do_map_with(xs, f, [map_f(f, acc, x) | acc], map_f)
   }
 }
 
-pub fn map(list, fun) {
-  do_map(list, fun, [])
+pub fn map(list, f) {
+  do_map_with(list, f, [], fn(fun, _, x) { fun(x) })
+}
+
+pub fn index_map(list, f) {
+  do_map_with(list, f, [], fn(fun, a, x) { fun(length(a), x) })
```

Comment:
The Gleam and Erlang compilers are not clever enough to optimise the call to `length` so the accumulator will be traversed on each iteration making this `O(n log n)` I'm afraid. We could get it down to `O(n)` using an implementation something like this:

```rust
fn do_index_map(list, fun, index, acc) {
  case list {
  | [] -> reverse(acc)
  | [x | xs] -> do_index_map(xs, fun, index + 1, [fun(index, x) | acc])
  }
}

pub fn index_map(list, fun) {
  do_index_map(list, fun, 0, [])
}
```